src/crimson/os/seastore/lba_manager/btree/btree_lba_manager.cc

```diff
@@ -319,76 +317,94 @@ BtreeLBAManager::_alloc_extent(
+	return LBABtree::iterate_repeat(
+	  c,
+	  btree.upper_bound_right(c, hint),
+	  [this, &state, total_len, addr, &t, hint, lookup_attempts](auto &pos) {
+	    LOG_PREFIX(BtreeLBAManager::_alloc_extent);
+	    ++stats.num_alloc_extents_iter_nexts;
+	    if (pos.is_end()) {
+	      DEBUGT("{}~{}, hint={}, state: end, done with {} attempts, insert at {}",
+		     t, addr, total_len, hint,
+		     stats.num_alloc_extents_iter_nexts - lookup_attempts,
+		     state.last_end);
+	      state.insert_iter = pos;
+	      return typename LBABtree::iterate_repeat_ret_inner(
+		interruptible::ready_future_marker{},
+		seastar::stop_iteration::yes);
+	    } else if (pos.get_key() >= (state.last_end + total_len)) {
+	      DEBUGT("{}~{}, hint={}, state: {}~{}, done with {} attempts, insert at {} -- {}",
+		     t, addr, total_len, hint,
+		     pos.get_key(), pos.get_val().len,
+		     stats.num_alloc_extents_iter_nexts - lookup_attempts,
+		     state.last_end,
+		     pos.get_val());
+	      state.insert_iter = pos;
+	      return typename LBABtree::iterate_repeat_ret_inner(
+		interruptible::ready_future_marker{},
+		seastar::stop_iteration::yes);
+	    } else {
+	      state.last_end = pos.get_key() + pos.get_val().len;
+	      TRACET("{}~{}, hint={}, state: {}~{}, repeat ... -- {}",
+		     t, addr, total_len, hint,
+		     pos.get_key(), pos.get_val().len,
+		     pos.get_val());
+	      return typename LBABtree::iterate_repeat_ret_inner(
+		interruptible::ready_future_marker{},
+		seastar::stop_iteration::no);
 	    }
-	    ceph_assert(inserted);
-	    state.ret = iter;
+	  }).si_then([c, addr, hint, &btree, &state, &alloc_infos, total_len, &rets] {
+	    return trans_intr::do_for_each(
+	      alloc_infos,
+	      [c, addr, hint, &btree, &state, total_len, &rets](auto &alloc_info) {
+	      return btree.insert(
+		c,
+		*state.insert_iter,
+		state.last_end,
+		lba_map_val_t{
+		  alloc_info.len,
+		  pladdr_t(alloc_info.val),
+		  1,
+		  alloc_info.checksum},
+		alloc_info.extent
+	      ).si_then([&state, c, addr, total_len, hint,
+			&alloc_info, &rets](auto &&p) {
+		LOG_PREFIX(BtreeLBAManager::_alloc_extent);
+		auto [iter, inserted] = std::move(p);
+		TRACET("{}~{}, hint={}, inserted at {}",
+		       c.trans, addr, total_len, hint, state.last_end);
+		if (alloc_info.extent) {
+		  ceph_assert(alloc_info.val.is_paddr());
+		  alloc_info.extent->set_laddr(iter.get_key());
+		}
+		ceph_assert(inserted);
+		rets.emplace_back(iter.get_pin(c));
+		return iter.next(c).si_then([&state, &alloc_info](auto it) {
+		  state.insert_iter = it;
+		  state.last_end += alloc_info.len;
+		});
+	      });
+	    });
 	  });
-	});
-    }).si_then([c](auto &&state) {
-      return alloc_extent_iertr::make_ready_future<
-	LBAMappingRef>(state.ret->get_pin(c));
-    });
+      }).si_then([&rets](auto &&state) {
+	return alloc_extent_iertr::make_ready_future<
+	  std::vector<LBAMappingRef>>(std::move(rets));
+      });
+  });
```

Comment:
Please elaborate in the commit message why the above chances are necessary.

Reply:
OK, seems to me an important reason to introduce `_alloc_extents()` instead of reusing ` _alloc_extent()`  is for the better performance.